compiler/noirc_frontend/src/elaborator/mod.rs

```diff
@@ -779,4 +796,310 @@ impl<'context> Elaborator<'context> {
+                        };
+                        self.push_err(error);
+                    }
+                }
+            }
+        }
+    }
+
+    fn elaborate_trait_impl(&mut self, trait_impl: UnresolvedTraitImpl) {
+        self.file = trait_impl.file_id;
+        self.local_module = trait_impl.module_id;
+
+        let unresolved_type = trait_impl.object_type;
+        let self_type_span = unresolved_type.span;
+        self.add_generics(&trait_impl.generics);
+
+        let trait_generics =
+            vecmap(&trait_impl.trait_generics, |generic| self.resolve_type(generic.clone()));
+
+        let self_type = self.resolve_type(unresolved_type.clone());
+        let impl_id = self.interner.next_trait_impl_id();
+
+        self.self_type = Some(self_type.clone());
+        self.current_trait_impl = Some(impl_id);
+
+        let mut methods = trait_impl.methods.function_ids();
+
+        self.elaborate_functions(trait_impl.methods);
+
+        let maybe_trait_id = trait_impl.trait_id;
+        if let Some(trait_id) = maybe_trait_id {
+            for func_id in &methods {
+                self.interner.set_function_trait(*func_id, self_type.clone(), trait_id);
+            }
+        }
+
+        if matches!(self_type, Type::MutableReference(_)) {
+            let span = self_type_span.unwrap_or_else(|| trait_impl.trait_path.span());
+            self.push_err(DefCollectorErrorKind::MutableReferenceInTraitImpl { span });
+        }
+
+        if let Some(trait_id) = maybe_trait_id {
+            let where_clause = trait_impl
+                .where_clause
+                .into_iter()
+                .flat_map(|item| self.resolve_trait_constraint(item))
+                .collect();
+
+            let resolved_trait_impl = Shared::new(TraitImpl {
+                ident: trait_impl.trait_path.last_segment().clone(),
+                typ: self_type.clone(),
+                trait_id,
+                trait_generics: trait_generics.clone(),
+                file: trait_impl.file_id,
+                where_clause,
+                methods,
+            });
+
+            let generics = vecmap(&self.generics, |(_, type_variable, _)| type_variable.clone());
+
+            if let Err((prev_span, prev_file)) = self.interner.add_trait_implementation(
+                self_type.clone(),
+                trait_id,
+                trait_generics,
+                impl_id,
+                generics,
+                resolved_trait_impl,
+            ) {
+                self.push_err(DefCollectorErrorKind::OverlappingImpl {
+                    typ: self_type.clone(),
+                    span: self_type_span.unwrap_or_else(|| trait_impl.trait_path.span()),
+                });
+
+                // The 'previous impl defined here' note must be a separate error currently
+                // since it may be in a different file and all errors have the same file id.
+                self.file = prev_file;
```

Comment:
Something like a `defer` would be useful but we couldn't capture `self` with it anyway. I think adding a wrapper function would obfuscate this too much since it'd be used in every function and we'd need several different versions for setting file, local module, and generics. Luckily most of these functions don't have any Results to handle.